pkg/cloud/rgraph/exec/executor_parallel.go

```diff
@@ -0,0 +1,189 @@
+/*
+Copyright 2024 Google LLC
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+https://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package exec
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"sync"
+	"time"
+
+	"github.com/GoogleCloudPlatform/k8s-cloud-provider/pkg/cloud"
+	"github.com/GoogleCloudPlatform/k8s-cloud-provider/pkg/cloud/rgraph/algo"
+	"k8s.io/klog/v2"
+)
+
+var (
+	ErrPendingActions = errors.New("Executor did not process all actions")
+)
+
+func defaultParallelExecutorConfig() *ExecutorConfig {
+	return &ExecutorConfig{
+		DryRun:                false,
+		ErrorStrategy:         ContinueOnError,
+		Timeout:               5 * time.Minute,
+		WaitForOrphansTimeout: 1 * time.Minute,
+	}
+}
+
+// NewParallelExecutor returns a new Executor that runs tasks multi-threaded.
+func NewParallelExecutor(c cloud.Cloud, pending []Action, opts ...Option) (*parallelExecutor, error) {
+	ret := &parallelExecutor{
+		config: defaultParallelExecutorConfig(),
+		cloud:  c,
+		result: &Result{Pending: pending},
+		pq:     algo.NewParallelQueue[Action](),
+	}
+	for _, opt := range opts {
+		opt(ret.config)
+	}
+
+	if err := ret.config.validate(); err != nil {
+		return nil, err
+	}
+	return ret, nil
+}
+
+type parallelExecutor struct {
+	config *ExecutorConfig
+	cloud  cloud.Cloud
+
+	// lock guards results
+	lock   sync.Mutex
+	result *Result
+
+	pq   *algo.ParallelQueue[Action]
+	done chan *TraceEntry
+}
+
+// parallelExecutor implements Executor.
+var _ Executor = (*parallelExecutor)(nil)
+
+// Run executes pending actions in parallel.
+//
+// ParallelExecutor will stop execution (to the extent possible) if the context
+// passed to Run() is cancelled. This will also cancel any waiting for orphan go
+// routines that are currently executing.
+//
+// To handle timeout properly use Executor Config. Set TimeoutOption for
```

Comment:
ExecutorConfig is not something accessible to the user and should not be in the public docs. You should only reference the public API that you want users to use, not anything package private.